src/main/java/dev/openfeature/sdk/EventEmitter.java

```diff
@@ -0,0 +1,108 @@
+        handlerMap.put(ProviderEvent.PROVIDER_READY, new ArrayList<>());
+        handlerMap.put(ProviderEvent.PROVIDER_CONFIGURATION_CHANGED, new ArrayList<>());
+        handlerMap.put(ProviderEvent.PROVIDER_ERROR, new ArrayList<>());
+        handlerMap.put(ProviderEvent.PROVIDER_STALE, new ArrayList<>());
+    }
+
+    /**
+     * Emit an event.
+     *
+     * @param event   Event type to emit.
+     * @param details Event details.
+     */
+    public void emit(ProviderEvent event, ProviderEventDetails details) {
+        try (AutoCloseableLock __ = handlersLock.readLockAutoCloseable()) {
+            EventDetails eventDetails = EventDetails.builder()
+                    .eventMetadata(details.getEventMetadata())
+                    .flagsChanged(details.getFlagsChanged())
+                    .message(details.getMessage())
+                    .build();
+
+            // we may be forwarding this event, preserve the name if so.
+            if (EventDetails.class.isInstance(details)) {
+                eventDetails.setClientName(((EventDetails) details).getClientName());
+            }
+
+            this.handlerMap.get(event).stream().forEach(handler -> {
+                runHander(handler, eventDetails);
+            });
+        }
+    }
+
+    void runHander(Consumer<EventDetails> handler, EventDetails eventDetails) {
+        taskExecutor.submit(() -> {
+            try {
+                handler.accept(eventDetails);
+            } catch (Exception e) {
+                log.error("Exception in event handler {}", handler, e);
+            }
+        });
+    }
+
+    void addHandler(ProviderEvent event, Consumer<EventDetails> handler) {
+        try (AutoCloseableLock __ = handlersLock.writeLockAutoCloseable()) {
+            this.handlerMap.get(event).add(handler);
+        }
+    }
+
+    void removeHandler(ProviderEvent event, Consumer<EventDetails> handler) {
+        try (AutoCloseableLock __ = handlersLock.writeLockAutoCloseable()) {
+            this.handlerMap.get(event).remove(handler);
+        }
+    }
+
+    void removeAllHandlers() {
+        try (AutoCloseableLock __ = handlersLock.writeLockAutoCloseable()) {
+            this.handlerMap.keySet().stream()
+                .forEach(type -> handlerMap.get(type).clear());
+        }
+    }
+
+    /**
+     * Propagates all events from the originatingEmitter to this one.
+     *
+     * @param originatingEmitter The emitter to forward events from.
+     * @param clientName         The client name that will be added to the events.
+     */
+    void forwardEvents(EventEmitter originatingEmitter, @Nullable String clientName) {
+        Arrays.asList(ProviderEvent.values()).stream().forEach(eventType -> {
+            originatingEmitter.addHandler(eventType, details -> {
+                // set the client name when we proxy the events through.
+                details.setClientName(clientName);
+                this.emit(eventType, details);
```

Comment:
Any reason to emit event when forwarding handlers 🤔  ?

Reply:
Events are forwarded _by_ emitting them again.

Basically, forwarding amounts to listening to events from the original emitter, and re-emitting them.

Reply:
I am still confuesed on this. For example, let's take the following scenario - setting the default provider. Current call stack looks like below,

API.setProvider(provider) - > API.propagateEventsIfSupported(provider, clientName) -> emitter.forwardEvents(eventProvider, clientName) -> emitter.emit(eventType, details)

This flow suggests adding known API-level events to the new provider and then running them. Or am I getting it wrong here.

Reply:
I've completely reworked this.

Now all handlers are stored in a single map, instead of doing this complex proxying.